compiler/rustc_mir_transform/src/upvar_to_local_prop.rs

```diff
@@ -0,0 +1,694 @@
+//!
+//! * Furthermore, locals can occur in the projection itself, e.g. as `Index(_3)` in a PlaceElem.
+//!   We cannot replace such locals with the `_1.0`, so this should likewise make the local
+//!   ineligible.
+//!
+//! Notes around soundness:
+//!
+//! * any writes to a local with a rhs that isn't either the upvar or a local in that upvar's set
+//!   should cause that local and all locals derived *from it* to become ineligible.
+//!
+//! * any writes to `_1` should invalidate the whole transformation. (Should be at most a corner
+//!   case.)
+//!
+//! * any write to `_1.L` where `.L` is one of the upvars that we are using as root values should
+//!   invalidate the use of that upvar as a root value. e.g. a write to `_1.2` should make `_1.2`
+//!   ineligible for the transformation, but *other* upvars like `_1.0` should still be able to
+//!   participate.
+//!
+//! But, what about transmutes of `&_1` and writes to fields of the transmuted thing? We simplify
+//! the reasoning here by just treating *any* use of `_1` that isn't a trivial assignment of a
+//! field to a local (e.g. `&mut _1`, `&_1`) as invalidating the whole transformation
+
+use crate::MirPass;
+use rustc_index::vec::{Idx, IndexVec};
+use rustc_middle::mir::visit::{MutVisitor, PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_span::def_id::DefId;
+
+struct IndexMap<K: Idx, V> {
+    vec: IndexVec<K, Option<V>>,
+}
+
+impl<K: Idx, V: Copy> IndexMap<K, V> {
+    fn new(n: usize) -> Self {
+        IndexMap { vec: IndexVec::from_elem_n(None, n) }
+    }
+    fn get(&self, k: K) -> Option<V> {
+        self.vec.get(k).map(|p| *p).flatten()
+    }
+}
+
+const SELF_IDX: u32 = 1;
+const SELF_ARG: Local = Local::from_u32(SELF_IDX);
+
+pub struct UpvarToLocalProp;
+
+// This is a way to make `trace!` format strings shorter to placate tidy.
+const W: &'static str = "UpvarToLocalProp";
+
+struct Patch<'tcx> {
+    tcx: TyCtxt<'tcx>,
+
+    /// Maps each local to upvar that transformation will use as local's replacement
+    local_to_root_upvar_and_ty: IndexMap<Local, (Field, Ty<'tcx>)>,
+}
+
+impl<'tcx> MirPass<'tcx> for UpvarToLocalProp {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        match sess.opts.unstable_opts.mir_upvar_to_local_prop {
+            Some(flag) => flag,
+            None => sess.mir_opt_level() > 0, // on by default w/o -Zmir-opt-level=0
+        }
+    }
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        trace!("Running UpvarToLocalProp on {:?}", body.source);
+
+        // First: how many upvars might we need to consider?
+        let upvar_limit: usize = {
+            // If we are looking at a generator, we have a fast path
+            if body.yield_ty().is_some() {
+                // The first argument is the generator type passed by value
+                let gen_ty = body.local_decls.raw[SELF_IDX as usize].ty;
```

Comment:
This may need to peel refs to fetch the actual generator type.

Reply:
you mean because the generator might be receiving a `self` type not by value, right?